target_chains/solana/programs/pyth-solana-receiver/src/lib.rs

```diff
@@ -368,61 +438,159 @@ fn post_price_update_from_vaa<'info>(
     vaa_payload: &[u8],
     price_update: &MerklePriceUpdate,
 ) -> Result<()> {
+    pay_single_update_fee(config, treasury, payer)?;
+    verify_vaa_data_source(config, vaa_components)?;
+    let message = verify_merkle_proof(vaa_payload, price_update)?;
+    match message {
+        Message::PriceFeedMessage(price_feed_message) => {
+            price_update_account.write_authority = write_authority.key();
+            price_update_account.verification_level = vaa_components.verification_level;
+            price_update_account.price_message = price_feed_message;
+            price_update_account.posted_slot = Clock::get()?.slot;
+        }
+        Message::TwapMessage(_) | Message::PublisherStakeCapsMessage(_) => {
+            return err!(ReceiverError::UnsupportedMessageType);
+        }
+    }
+    Ok(())
+}
+
+#[allow(clippy::too_many_arguments)]
+fn post_twap_update_from_vaas<'info>(
+    config: &Account<'info, Config>,
+    payer: &Signer<'info>,
+    write_authority: &Signer<'info>,
+    treasury: &AccountInfo<'info>,
+    twap_update_account: &mut Account<'_, TwapUpdate>,
+    start_vaa_components: &VaaComponents,
+    end_vaa_components: &VaaComponents,
+    start_vaa_payload: &[u8],
+    end_vaa_payload: &[u8],
+    start_price_update: &MerklePriceUpdate,
+    end_price_update: &MerklePriceUpdate,
+) -> Result<()> {
+    pay_single_update_fee(config, treasury, payer)?;
+
+    // Verify data sources for both VAAs
+    for vaa_components in [start_vaa_components, end_vaa_components] {
+        verify_vaa_data_source(config, vaa_components)?;
+    }
+
+    // Verify both merkle proofs and extract their messages
+    let start_message = verify_merkle_proof(start_vaa_payload, start_price_update)?;
+    let end_message = verify_merkle_proof(end_vaa_payload, end_price_update)?;
+
+    // Calculate the TWAP and store it in the output account
+    match (start_message, end_message) {
+        (Message::TwapMessage(start_msg), Message::TwapMessage(end_msg)) => {
+            let (price, conf, down_slots_ratio) = calculate_twap(&start_msg, &end_msg)?;
+
+            twap_update_account.write_authority = write_authority.key();
+            twap_update_account.verification_level = start_vaa_components.verification_level;
+
+            twap_update_account.twap.feed_id = start_msg.feed_id;
```

Comment:
I think you forgot to check that start and end are for the same id?

Reply:
yea true, will update